src/iterative_ensemble_smoother/_iterative_ensemble_smoother.py

```diff
@@ -88,66 +130,76 @@ def fit(
-                mean=np.zeros_like(observation_values),
-                cov=observation_errors,
-                size=ensemble_size,
-                method="cholesky",  # An order of magnitude faster than 'svd'
-            ).T
-        # A vector of standard deviations was passed
-        else:
-            E = rng.normal(
-                loc=0, scale=observation_errors, size=(ensemble_size, num_obs)
-            ).T
+        # ---------------------------------------------------------------------
+        # ----------- Computations corresponding to algorithm 1 ---------------
+        # ---------------------------------------------------------------------
 
+        # Draw samples from N(0, C_dd)
+        E = self._get_E(
+            observation_errors=observation_errors,
+            observation_values=observation_values,
+            ensemble_size=ensemble_size,
+        )
         assert E.shape == (num_obs, ensemble_size)
 
-        E -= E.mean(axis=1, keepdims=True)
-
-        R, observation_errors_std = _create_errors(observation_errors, inversion)
+        # Transform covariance matrix to correlation matrix
+        R, observation_errors_std = covariance_to_correlation(observation_errors)
 
         # Store D as defined by Equation (14) in Evensen (2019)
         self.D_ = E + observation_values.reshape(num_obs, 1)
         D = self.D_ - response_ensemble
 
+        # Note on scaling
+        # -------------------
+        # In line 8 in Algorithm 1, we solve (S S^T + E E^T) X = (SW + D - g(X))
+        # for X, and then we compute S.T X. If we scale the rows (observed variables)
+        # of these equations using the standard deviations, we can obtain better
+        # conditioning numbers on the equation. This corresponds to left-multiplying
+        # with a diagonal matrix L := sqrt(diag(C_dd)). In an experiment with
+        # random covariance matrices, this improved the condition number ~90%
+        # of the time.
+        # To see the equality, note that if we scale S, E and K := (SW + D - g(X))
+        # we obtain
+        #              (L S) (L S)^T + (L E) (L E)^T X_2 = L K
+        #              L (S S^T + E E^T) L X_2 = L K
+        # so the new solution X_2 := L^-1 X, expressed in terms of the original X.
+        # But when we left-multiply with S^T, we do so with a transformed S,
+        # so we obtain S_2^T X_2 = (L S)^T (L^-1 X) = S^T X, so the solution
+        # to the transformed system is equal to the solution of the original system.
+        # In the implementation of scaling the right hand side (SW + D - g(X))
+        # we first scale D - g(X), then we scale S implicitly by solving
+        # S Sigma = L Y, instead of S Sigma = Y for S.
+
         # Scale D and E with observation error standard deviations.
         D /= observation_errors_std.reshape(num_obs, 1)
-        E /= observation_errors_std.reshape(num_obs, 1)
+        E = E / observation_errors_std.reshape(num_obs, 1)
```

Comment:
From my understanding, this creates a copy of `E`, while using the in-place operator `/=` does not. Am I wrong?

Reply:
You are correct! We have to copy here, since we do not draw E again and again in each iteration now. We use the same E, and then we cannot scale in each iteration.

This is a bit messy overall, I hope to fix more of it in future PRs - but for now the memory test still passes, so this approach does not use much more memory

Reply:
Not sure I follow.
Both `E = E / observation_errors_std.reshape(num_obs, 1)` and `E /= observation_errors_std.reshape(num_obs, 1)` should produce the same result but the last one will use less memory right?

Reply:
If we don't cache E, then we have to cholesky decompose the covariance matrix in every iteration, plus multiply the cholesky factor with a vector. That's not nothing, but I have not timed it. Certainly more time and memory than if we cache it (even with mutating it).

In the future it will not be mutated. E is used for 2 purposes in the algorithm: (1) to add noise to observations and (2) to create a low-rank approximation of the covariance. Actually, there is no reason why E has to have the same dimensional for these purposes. If we have 5 output variables and 2 ensemble members, then we use E of shape (5, 2) for both purposes. But using E of shape e.g. (5, 3) would be a better approximation to the covariance matrix. We can separate this clearly and document it.

In the future I do not want the code to look like it does in this PR, but I am trying to attack the problem step by step. Caching E removes some unneccessary work (the repeated cholesky), but there is still work to do

Reply:
I'm principally against adding any kind of state unless it is of considerable benefit, and I am not sure whether that is the case here.
I've not seen E used to create a low-rank approximation in the wild so I do not deem it very important.
Let's have a think and discuss via video.